[Refactor] Adopt request builder patterns for SecurityRestApiActions for consistency and clarity

willyborankin opened the pull request:
### Description

REST API clean up continuation.

Main differences are: 
- Using functional approach to handle requests instead of inheritance  which (IMHO) simplify code support and reading 
- All checks and verification stay the same, I only changed names. 
- PATCH uses the same validation rules as PUT and DELETE methods 


### Check List
- [ ] New functionality includes testing
- [ ] New functionality has been documented
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

willyborankin commented:
@peternied and @cwperks as i promised. So far all methods aka CRUD. PATCH will be added a bit later.

willyborankin commented:
I fixed current failures

peternied reviewed (commented):
Added some comments around this draft - overall a big fan of this direction.

willyborankin commented on build.gradle:
```diff
@@ -92,6 +92,12 @@ spotbugsTest {
 
+spotless {
+    java {
+        toggleOffOn()
```
well e.g. for InternalUserApi it looks like that:
```java
    @Override
    protected void configureRequestHandlers(RequestHandler.RequestHandlersBuilder requestHandlersBuilder) {
        // Overrides the GET request functionality to allow for the special case of requesting an auth token.
        requestHandlersBuilder.override(
            Method.POST,
            (channel, request, client) -> withAuthTokenPath(request).map(this::loadFilteredConfiguration)
                .map(this::resourceExists)
                .valid(securityConfiguration -> generateAuthToken(channel, securityConfiguration))
                .error((status, toXContent) -> Responses.response(channel, status, toXContent))
        )
            .onChangeRequest(
                Method.PUT,
                request -> processPutRequest(request).map(securityConfiguration -> createOrUpdateUser(request, securityConfiguration))
            );
    }
```
vs.
```java
    @Override
    protected void configureRequestHandlers(RequestHandler.RequestHandlersBuilder requestHandlersBuilder) {
        // spotless:off
        // Overrides the GET request functionality to allow for the special case of requesting an auth token.
        requestHandlersBuilder
                .override(Method.POST, (channel, request, client) ->
                        withAuthTokenPath(request)
                                .map(this::loadFilteredConfiguration)
                                .map(this::resourceExists)
                                .valid(securityConfiguration -> generateAuthToken(channel, securityConfiguration))
                                .error((status, toXContent) -> Responses.response(channel, status, toXContent)))
                .onChangeRequest(Method.PUT, request ->
                        processPutRequest(request)
                                .map(securityConfiguration -> createOrUpdateUser(request, securityConfiguration))
        );
        // spotless:on
    }
```

TBH, I added toggleOffOn in the hope that somebody will give a hint on how to fix it. :-D

willyborankin replied:
True I read the same and do not want to change formatter especially after @scrawfor99 reformatted the whole codebase.

peternied replied:
> we can discuss during the next triage meeting

That is a great idea - look forward to it.

willyborankin replied:
removed

willyborankin replied:
@peternied I marked it as resolved since I removed spotles:on/off from the build.gradle file

willyborankin commented on src/main/java/org/opensearch/security/dlic/rest/api/AbstractApiAction.java:
```diff
@@ -113,150 +128,208 @@ protected AbstractApiAction(
+            .onGetRequest(this::processGetRequest)
+            .onChangeRequest(Method.DELETE, this::processDeleteRequest)
+            .onChangeRequest(Method.PUT, this::processPutRequest);
+        configureRequestHandlers(requestHandlersBuilder);
```
For such solution, I have to define the builder as a protected final variable and use it on the constructor level in `AbstractAPI `sub- classes. Personally, I do not like such things, but if you think that it is ok i will do it no problem. But I suspect that I'm missing something :-)

peternied replied:
I think making the builder internal to `org.opensearch.security.dlic.rest.api` is good enough far as protections go.

willyborankin replied:
@peternied I moved `configureRequestHandlers` into `RequestHandlersBuilder`

cwperks commented on src/main/java/org/opensearch/security/dlic/rest/api/AbstractApiAction.java:
```diff
@@ -113,150 +128,215 @@ protected AbstractApiAction(
-                getConfigName(),
-                existingConfiguration,
-                new OnSucessActionListener<IndexResponse>(channel) {
+    protected final ValidationResult<SecurityConfiguration> loadFilteredConfiguration(final String resourceName) throws IOException {
```
What does filtered mean here exactly? If I'm reading this correctly it looks like its removing hidden, clearing hashes and setting meta to null?

In some instances I know `filter` is used to search through the list of roles for roles that match a given name. Is this method used for that? Example: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java#L186-L188

willyborankin replied:
> What does filtered mean here exactly? If I'm reading this correctly it looks like its removing hidden, clearing hashes and setting meta to null?
> 

Yes it is and I think the name is misleading. What do you think to call it something like:
`loadWithoutSensitiveData` or something else. Im not good in naming :-)

willyborankin replied:
Done

willyborankin replied:
done

cwperks commented on src/main/java/org/opensearch/security/dlic/rest/api/SecurityConfiguration.java:
```diff
@@ -0,0 +1,121 @@
+/*
```
Is it possible to add more tests around this class? Could `resourceName` ever be `null` in this class?

cwperks replied:
Does this class represent a single entry in the security index? i.e. A single internal user or a single role?

willyborankin replied:
I will look to it one more time, may I can reed of it. I added it since java compiler in JDK 11 can''t resolve `SecurityDynamicConfiguration<?>` type if you use it with `Optional<>` or `ValidationResult<>`

willyborankin replied:
I made `SecurityConfiguration` as a companion class.

willyborankin replied:
Made `SecurityConfiguration` as a companion class.

cwperks commented on src/main/java/org/opensearch/security/dlic/rest/api/AbstractApiAction.java:
```diff
@@ -113,150 +128,215 @@ protected AbstractApiAction(
-        }
+    protected final ValidationResult<SecurityConfiguration> resourceCanBeChanged(final SecurityConfiguration securityConfiguration)
+        throws IOException {
+        return resourceNotHidden(securityConfiguration).map(ignore -> resourceNotReadOnly(securityConfiguration));
```
Do static or reserved need to be considered here?

willyborankin replied:
Both, but I will change the function body. There is no need to use chaining here.

willyborankin replied:
Address in the latest commit. 
All configuration validation extracted into a separate class: `ConfigurationValidators` together with the test `ConfigurationValidatorsTest`.

willyborankin replied:
Changed it.

cwperks reviewed (commented):
Thank you @willyborankin ! I took a first pass at a review for this PR and left a few comments. Overall, I like the introduction of higher level abstractions and creating a common class for common Responses.

willyborankin commented:
Renamed methods:
 - `getResourceName` -> `getConfigurationName`
 - `getConfigName` -> `getConfigurationType`

cwperks commented:
@willyborankin I got an email about a comment, but cannot see it any longer:

> Renamed methods:
> getResourceName -> getConfigurationName
> getConfigName -> getConfigurationType


I like the `getConfigurationType` (or `getConfigType`) name as it connects to [`CType`](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/securityconf/impl/CType.java) well.

Likewise `getConfigEntryByName` could work for `getResourceName`, but I am also fine with `getResourceName` as well. What should we call a single entry in one of the sections of the security config types (internal_users, roles, roles_mappings, etc.)? Should these generally be referred to as ConfigEntries or as resources?

willyborankin commented:
> @willyborankin I got an email about a comment, but cannot see it any longer:
> 
> > Renamed methods:
> > getResourceName -> getConfigurationName
> > getConfigName -> getConfigurationType
> 
> I like the `getConfigurationType` (or `getConfigType`) name as it connects to [`CType`](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/securityconf/impl/CType.java) well.
> 
> Likewise `getConfigEntryByName` could work for `getResourceName`, but I am also fine with `getResourceName` as well. What should we call a single entry in one of the sections of the security config types (internal_users, roles, roles_mappings, etc.)? Should these generally be referred to as ConfigEntries or as resources?

Yes I reverted found a bug. 
Names closer to your suggestion. `getConfigName` -> `getConfigType`. A signle entity name I use so far `entityName` :-) .

willyborankin commented:
Hi, @peternied and @cwperks.

I finished and address changes we discussed.

Changes:
 - added PATCH
 - removed all dead code
 - removed spotless on/off

Added a common `EndpointValidator` class which uses for all HTTP methods and validate configuration before it will be processed by HTTP response. So PATCH now uses the same set of validations rules as for PUT and DELETE methods.

cwperks commented on src/main/java/org/opensearch/security/dlic/rest/api/AbstractApiAction.java:
```diff
@@ -113,177 +138,319 @@ protected AbstractApiAction(
-            // throw jme;
-            // } else throw new JsonMappingException(null, jme.getMessage());
+            for (final var patchOperation : operations) {
+                if (!PATCH_OPERATIONS.contains(patchOperation)) {
```
Curious, should this be case insensitive? What is the current behavior?

willyborankin replied:
This is a good question. I use upper case for `final static` variables and it is a habit nothing more. Out check style tells nothing about it. I can change it to `PATCH_OPERATIONS` as `patchOpertaions`.

willyborankin replied:
@cwperks Looks like we use default rules about naming in case of final static:
https://github.com/opensearch-project/security/blob/32d3112ad3415243572928a70e64e382cbdfd43b/checkstyle/sun_checks.xml#L106C40-L106C40

which is: 
```java
Code Example:

 class MyClass {
   public static int goodStatic = 2; // OK
   private static int BadStatic = 2; // violation, name 'BadStatic'
                                     // must match pattern '^[a-z][a-zA-Z0-9]*$'
 }
```
Wondering why it passes check style :-)

cwperks commented on src/main/java/org/opensearch/security/dlic/rest/api/RolesApiAction.java:
```diff
@@ -42,9 +41,13 @@
-public class RolesApiAction extends PatchableResourceApiAction {
+public class RolesApiAction extends AbstractApiAction {
+
+    protected final static String RESOURCE_NAME = "role";
```
Is it possible to use the CType enum here? 

```
protected final static String RESOURCE_NAME = CType.ROLES.toLCString();
```

cwperks replied:
nevermind, I see that this is singular

willyborankin replied:
Ahhh yes I tried to use `CType.ROLES.toLCString()` as a resource name but the response in this case looke lime this: `roles 'aaaaa' not found`. I will remve it and move resource name to the validator.

cwperks reviewed (commented):
@willyborankin This PR looks good to me. I like the change this introduces to reduce the amount of `notImplemented` around the concrete APIs.

What work is remaining before marking this ready for review?

willyborankin commented:
> @willyborankin This PR looks good to me. I like the change this introduces to reduce the amount of `notImplemented` around the concrete APIs.
> 
> What work is remaining before marking this ready for review?

I'm going to add 1 test and open it as `ready for review`

peternied commented on src/main/java/org/opensearch/security/dlic/rest/api/AbstractApiAction.java:
```diff
@@ -113,177 +139,313 @@ protected AbstractApiAction(
+    }
 
-                }
+    protected EndpointValidator createEndpointValidator() {
```
Is this validator unused, seems like its has no code coverage?

willyborankin replied:
Yes due to the problem with mocking this class, but your idea with `AbstractApiActionDependencies` class "open gates" to do it :-)

peternied reviewed (commented):
Looking really good, I've reviewed almost everything and pending a couple of open issues we can see about getting this merged.

willyborankin commented on src/main/java/org/opensearch/security/dlic/rest/api/AbstractApiAction.java:
```diff
@@ -113,177 +138,319 @@ protected AbstractApiAction(
-        filter(configuration);
-        // no specific resource requested, return complete config
-        if (resourcename == null || resourcename.length() == 0) {
+    protected abstract String getResourceName();
```
Agree

willyborankin commented on src/main/java/org/opensearch/security/dlic/rest/api/SecurityRestApiActions.java:
```diff
@@ -67,79 +65,15 @@ public static Collection<RestHandler> getHandler(
-                threadPool,
-                auditLog
-            )
+            new RolesMappingApiAction(settings, configPath, adminDns, cr, cs, principalExtractor, evaluator, threadPool, auditLog)
```
Agree it gives possibility to use mocking for testing logic only.

peternied commented on src/main/java/org/opensearch/security/dlic/rest/api/RolesMappingApiAction.java:
```diff
@@ -119,21 +95,21 @@ private ValidationResult<SecurityConfiguration> validateRoleForMapping(final Sec
 
             @Override
-            public ValidationResult<SecurityConfiguration> hasRightsToChangeEntity(SecurityConfiguration securityConfiguration)
+            public ValidationResult<SecurityConfiguration> hasRightsToChangeImmutableEntity(SecurityConfiguration securityConfiguration)
```
Nit; hasRightsToChangeImmutableEntity -> isAllowedToChangeImmutableEntity

willyborankin replied:
yes looks much better

peternied commented on src/main/java/org/opensearch/security/OpenSearchSecurityPlugin.java:
```diff
@@ -238,6 +238,21 @@ public static void traceAction(String message, Object p0) {
+        }
+
+        AccessController.doPrivileged((PrivilegedAction<Object>) () -> {
+            if (Security.getProvider(BouncyCastleProvider.PROVIDER_NAME) == null) {
+                Security.addProvider(new BouncyCastleProvider());
+            }
```
Why move this?

peternied reviewed (changes requested):
I don't think we should change when bouncy castle is initialized, doesn't seem needed for this PR.  Looks like there might be some tests issues, but once CI is green, looking good to merge from me.

Note; For this latest commit there was a whole host of 'not covered' errors, I've created a PR to prevent these false positive scenarios in the future [1] 

- [1] https://github.com/opensearch-project/security/pull/3242

willyborankin commented:
> I don't think we should change when bouncy castle is initialized, doesn't seem needed for this PR. Looks like there might be some tests issues, but once CI is green, looking good to merge from me.
> 
> Note; For this latest commit there was a whole host of 'not covered' errors, I've created a PR to prevent these false positive scenarios in the future [1]
> 
>     * [1] [Retry code coverage upload on failure #3242](https://github.com/opensearch-project/security/pull/3242)

@peternied thank you that you spotted it. I was investigating this problem together with @reta and accidentally added this code to this branch. Sure must be removed and in any case it doesn't solve the problem.

willyborankin commented:
@peternied Thank you the coverage looks better now :-) 
```Successful in 1s — 83.82% of diff hit (target 62.49%) ```

peternied commented on src/main/java/org/opensearch/security/dlic/rest/api/InternalUsersApiAction.java:
```diff
@@ -111,221 +91,178 @@ public List<Route> routes() {
+        return loadConfiguration(CType.ROLES, false, false).map(rolesConfiguration -> {
+            final var contentAsNode = (ObjectNode) securityConfiguration.requestContent();
+            final var securityJsonNode = new SecurityJsonNode(contentAsNode);
+            // FIXME do we need to verify roles as well?
```
Looks liked there is this and a couple other FIXME's in this PR.  What do you want to do with these create issue to follow up or address before merging?

willyborankin replied:
@peternied Thank you,
I'm planning to address them after PR is merged as soon as possible. Regarding this FIXME i will remove it. I misunderstood this part, we have `security_roles`, `backend_roles`, etc., so we do not need validate other roles except `security_roles`.

peternied replied:
Having one place to store all future efforts as issues in github is my preference, but ain't no big deal for small items.

peternied reviewed (approved):
I'm comfortable with a fast follow for the FIXME remarks.  I've updated the PR title as I think it better encompasses the 'what' of this change.
